## 종류

![](https://user-images.githubusercontent.com/70433341/104199159-6d2f1200-546a-11eb-89bd-207c4d2c149f.png)

1. Error
 - 컴퓨터 하드웨어의 오동작 또는 고장으로 인해 응용프로그램에 이상이 생겼거나 JVM실행에 문제가 생겼을 경우 발생하는 것이다.
 - 프로세스에 영향을 준다.
 - 시스템 레벨에서 발생한다 (자바 프로그램 외의 오류)
 - VirtualMachineError, OutOfMemoryError, StackOverflowError

2. Exception
 - 컴퓨터의 에러가 아닌 사용자의 잘못된 조작 또는 개발자의 잘못된 코딩으로 인해 발생하는 프로그램 오류
 - 예외가 발생하면 프로그램이 종료가 된다는 것은 에러와 동일하지만 예외는 예외처리를 통해 프로그램을 종료되지 않고 정상적으로 작동되게 만들어 줄 수 있다.
 - 개발자가 구현한 로직에서 발생
 - 쓰레드에 영향을 준다.

### 모든 예외의 부모 클래스는 java.lang.Throwable 클래스이다.
 
## Checked Exception 과 Unchecked Exception

### Checked Exception
- 체크란, compiler 단에서 체크가 가능한 에러를 말하는 것이다. 이는 어찌보면 Runtime Exception과 상반된다고 할 수 있다. 이 체크된 에러들, 즉 에러가 발생할 수 있을 때에는 예외처리를 진행해줘야 하는데, 자바에서 강제적으로 try catch문을 작성하도록 하게된다.

### Uncheked Exception
- Unchecked Exception 이라 함은, 컴파일러 자체에서 예외를 찾아내지 못하는 걸 말한다. Checked Exception과 다르게 예외처리를 강제로 지정하지 않는 특징이 있다.

### Checked Exception과 UncheckedException으로 나눈 이유
 - 예외도 메소드의 시그니처와 번환값 만큼 중용한 공용 인터페이스이다. 메소드를 호출하는 부분은 해당 메소드에서 어떤 예외를 던질 수 있는지 반드시 알아야 한다. 그렇기 때문에 Checked Exception으로 해당 메소드가 발생할 수 있는 예외를 명시하도록 강제한다.
 - 하지만 Unchecked Exception, 즉 Runtime Exception은 예외를 명시하지 않아도 된다. Runtime Exception은 프로그램의 코드 문제로 발생하는 예외이다. 그렇기 때문에 클라이언트 쪽에서 복구하거나 대처할 수 있다고 예상할 수 없고, 어디서나 일어날 수 있기 때문에 메서드에 명시할 경우 프로그램의 명확성을 떨어뜨릴 수 있다.
 - 정리하면 클라이언트가 Exception을 적절하게 회복할 수 있으면 Checked Exception으로 만들고, 그렇지 않은 경우 Unchecked Exception으로 만드는 것이 좋다.

### Unchecked Exception에 대한 오해
-  Spring framework에서 Transaction 설정과 관련하여 Unchecked Exception에 대해 roll-back기능을 지원한다. 이것은 Spring framework의 transaction 설정이 제공하는 것이다. 순수 Java 언어에서 지원하는 기능이 아니다. Java가 제공하는 Unchecked Exception에는 roll-back 기능이 없다.

## throws
 - 예외가 발생 했을때 try-catch로 묶어주지 않아도 메소드를 호출한 메소드로 예외처리를 위임하는것이기 때문에 문제되지 않는다.
 - 메소드 선언부에 throws Exception 선언을 하게 되면 해당 메소드는 예외처리를 안해도 되지만 호출한 메소드에서는 반드시 try-catch로 예외처리를 해야한다.
 - catch 블록에서 예외를 throw할 경우에도 메소드 선언의 throws 구문에 해당 예외가 정의되어 있어야 한다.

## RuntimeException과 아닌것의 차이
 - RuntimeException은 Unchecked Exception이고 실행하지 않고 예외가 검출되는 것(RE 외)이 Checked Exception이다.

### RuntimeException
 - RuntimeException 클래스를 상속받는 자식 클래스들은 주로 치명적인 예외 상황을 발생시키지 않는 예외들로 구성
 - try-catch 사용보다는 프로그램을 작성하면서 예외가 발생하지 않도록 주의하는것이 좋다.
 - 예외 발생 시 트랜잭션 rollback한다.

### 그외 Exception
 - 치명적인 예외 상황을 발생시키기 때문에 반드시 예외처리를 해야만 한다.
 - 컴파일러는 예외가 발생할 가능성이 있는 구문에는 반드시 예외를 처리하도록 강제하고 있다.
 - 컴파일 단계에서 확인 가능
 - 예외 발생 시 트랜잭션 rollback하지 않는다.

## JAVA 예외처리 전략
 - 실행시에 발생할 확률이 매우 높은 경우에는 런타임 예외로 만드는것이 나을 수도 있다. 즉, 클래스 선언시 extends Exception 대신에 extends RuntimeException으로 선언하는것이다.
   이렇게 되면 해당 예외를 던지는(throws) 메소드를 사용하더라도 try-catch로 묶지 않아도 컴파일시에 예외가 발생하지 않는다. 하지만 이 경우에는 예외가 발생할 경우 해당 클래스를 호출하는 다른 클래스에서 예외를 처리하도록 구조적인 안전장치가 되어있어야만 한다. (안전장치란 메소드를 호출하는 메소드에서 예외를 처리하는 것)
 - 임의의 커스텀 exception을 만들때에는 반드시 try-catch로 묶어줄 필요가 있을 경우에만 Exception 클래스를 확장한다. 일반적으로 실행시 예외를 처리할 수 있는 경우에는 RuntimeException 클래스를 확장하는것을 권장한다.
 - catch문 내에 아무런 작업없이 공백을 놔두면 예외 분석이 어려워지므로 로그처리와 같은 예외처리를 해줘야 한다. 
 - 대부분의 코드는 커스텀 예외를 던지기 전에 표준 예외를 캐치하는 케이스가 많다. 보통 캐치된 예외에는 오류를 분석하는데 필요한 중요 정보가 포함되어 있다.  그렇기 때문에 원인을 가지고 있는 Throwable을 받을 수 있는 생성자 메소드를 제공해야 한다. 발생한 Throwable을 파라미터를 통해 가져올 수 있는 생성자를 최소한 하나를 구현하고 super클래스에 Throwable을 전달해줘야 한다.

```java
public class MyBusinessException extends Exception {
    private static final long serialVersionUID = 7718828512143293558 L;
    private final ErrorCode code;
    public MyBusinessException(ErrorCode code) {
        super();
        this.code = code;
    }
    public MyBusinessException(String message, Throwable cause, ErrorCode code) {
        super(message, cause);
        this.code = code;
    }
    public MyBusinessException(String message, ErrorCode code) {
        super(message);
        this.code = code;
    }
    public MyBusinessException(Throwable cause, ErrorCode code) {
        super(cause);
        this.code = code;
    }
    public ErrorCode getCode() {
        return this.code;
    }
}
```
```java
public void handleExceptionInOneBlock() {
    try {
        wrapException(new String("99999999"));
    } catch (MyBusinessException e) {
        // handle exception
        log.error(e);
    }
}
private void wrapException(String input) throws MyBusinessException {
    try {
        // do something
    } catch (NumberFormatException e) {
        throw new MyBusinessException("A message that describes the error.", e, ErrorCode.INVALID_PORT_CONFIGURATION);
    }
}
```
